### PR TITLE
feat(policy): create MeshRateLimit api and validator

### PR DESCRIPTION
**Pull request by lukidzi**
- [X] Link to docs PR or issue --
- [X] Link to UI issue or PR --
- [X] Is the [issue worked on linked][1]? -- Part of: https://github.com/kumahq/kuma/issues/5361
- [X] The PR does not hardcode values that might break projects that depend on kuma (e.g. "kumahq" as a image registry) --
- [X] The PR will work for both Linux and Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [X] Unit Tests --
- [X] E2E Tests --
- [X] Manual Universal Tests --
- [X] Manual Kubernetes Tests --
- [X] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [X] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? --
- [X] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch) or add a `ci/` label to run fewer/more tests?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

> Changelog: feat(policy): implement MeshRateLimit policy

### CODE REVIEW

File: pkg/plugins/policies/meshratelimit/api/v1alpha1/meshratelimit.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	// Define if rate limiting should be disabled.
+	// Default: false
+	// +optional
+	Disabled bool `json:"disabled,omitempty"`
+
+	// The number of HTTP requests this RateLimiter allows
+	// +required
+	Requests uint32 `json:"requests,omitempty"`
+
+	// The the interval for which `requests` will be accounted.
```
Comment (lahabana) [inline pkg/plugins/policies/meshratelimit/api/v1alpha1/meshratelimit.go:51]:
```suggestion
	// The interval for which `requests` will be accounted.
```

File: pkg/plugins/policies/meshratelimit/api/v1alpha1/validator.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+			SupportedKinds: supportedKinds,
+		}))
+	}
+	return verr
+}
+
+func validateDefault(path validators.PathBuilder, conf Conf) validators.ValidationError {
+	var verr validators.ValidationError
+	local := conf.Local
+	if local == nil {
```
Comment (lahabana) [inline pkg/plugins/policies/meshratelimit/api/v1alpha1/validator.go:66]: In practice you could make `local` non pointer. I think it'd make this a little simpler

File: pkg/plugins/policies/meshratelimit/api/v1alpha1/validator.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+		return verr
+	}
+	path = path.Field("local")
+	if local.HTTP != nil {
+		verr.Add(validateLocalHttp(path.Field("http"), local.HTTP))
+	}
+	if local.TCP != nil {
+		verr.Add(validateLocalTcp(path.Field("tcp"), local.TCP))
+	}
```
Comment (lahabana) [inline pkg/plugins/policies/meshratelimit/api/v1alpha1/validator.go:75]: HTTP and TCP are not exclusive?

File: pkg/plugins/policies/meshratelimit/api/v1alpha1/validator_test.go [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+- targetRef:
+    kind: Mesh
+  default:
+    local:
+      tcp:
+        disabled: false`,
+				expected: `
+violations:
+  - field: spec.from[0].default.local.tcp.connections
+    message: must be greater than
```
Comment (lahabana) [inline pkg/plugins/policies/meshratelimit/api/v1alpha1/validator_test.go:188]: This is strange no?

File: pkg/plugins/policies/meshratelimit/api/v1alpha1/validator_test.go [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+- targetRef:
+    kind: Mesh
+  default:
+    local:
+      http:
+        disabled: false`,
+				expected: `
+violations:
+  - field: spec.from[0].default.local.http.requests
+    message: must be greater than
```
Comment (lahabana) [inline pkg/plugins/policies/meshratelimit/api/v1alpha1/validator_test.go:170]: Strange sentence

File: pkg/plugins/policies/meshratelimit/api/v1alpha1/validator_test.go [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+  default:
+    local:
+      http:
+        interval: 49ms
+      tcp:
+        interval: 49ms`,
+				expected: `
+violations:
+  - field: spec.from[0].default.local.http.requests
+    message: must be greater than
```
Comment (lahabana) [inline pkg/plugins/policies/meshratelimit/api/v1alpha1/validator_test.go:208]: same here

File: pkg/plugins/policies/meshratelimit/api/v1alpha1/validator_test.go [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+  - field: spec.from[0].default.local.http.interval
+    message: 'must be greater than: 50ms'
+  - field: spec.from[0].default.local.tcp.connections
+    message: must be greater than
```
Comment (lahabana) [inline pkg/plugins/policies/meshratelimit/api/v1alpha1/validator_test.go:212]: and here

**Review by lahabana — APPROVED** [review]
Some nits but LGTM

File: pkg/plugins/policies/meshratelimit/api/v1alpha1/validator.go (hunk 3 above)
Comment (lukidzi) [reply pkg/plugins/policies/meshratelimit/api/v1alpha1/validator.go:75]: We agreed that we can configure both: https://github.com/kumahq/kuma/pull/5333#discussion_r1028064886

File: pkg/plugins/policies/meshratelimit/api/v1alpha1/validator_test.go (hunk 4 above)
Comment (lukidzi) [reply pkg/plugins/policies/meshratelimit/api/v1alpha1/validator_test.go]: Forgot to set the value. Fixed

**Comment by lahabana** [comment]
@lukidzi what's missing to get this in?

**Pull request merged**